README.md
# Anubis 𓁢

Ask Anubis to watch your front-end assets. When those files change, Anubis will inject them and remove the dead link, or reload the browser altogether.

<img src="screenshot.png" width="458" alt="Screenshot of Anubis CLI">

## What Anubis is not

- [HMR](https://webpack.js.org/concepts/hot-module-replacement/) &mdash; this is not Hot Module Replacement and is not intended to replace it. Anubis is best used for websites with minimal client-managed state, and is not recommended for web apps or SPAs.
- [LiveReload](http://livereload.com/) &mdash; shares some ideas from the days of old, but is not at feature parity with LiveReload with things like injecting `<style>` and `<img>` tags.
- [BrowserSync](https://www.browsersync.io/) &mdash; shares some ideas with BrowserSync, but will never meet feature parity. Anubis does not sync scroll position, have a dashboard, etc.

## What Anubis excels at
- Injecting CSS changes (via `link` tags, no `style` support yet)
- Reloading the browser when other files change
- Being very fast and light weight
- Makes working with backend servers easy (does not expect your project to be an SPA)

## Usage

### Install
```shell
npm i -D https://github.com/billcolumbia/anubis
```

### Add npm script
```json
{
    "scripts": {
        "anubis": "anubis -f './public/**/*.{js,css,php}'"
    }
}
```

To use an array of globs for the files option, pass the same `-f` argument *n* times:

```shell
anubis -f './public/**/*.{js,css}' -f './templates/**/*.{twig,html}'
```

### Connect to the proxy
Once anubis is running make sure to connect to localhost at the port you provided (`localhost:3000` by default). If you go to the proxied URL, you wont get any injection or reloading!

## Options/API
```shell
npx anubis --help
```

### `port`:     
**about**: Port to proxy

**required**: no

**type**: number

**default**: 3000

### `target`: 
**about**: URL including port (root for your backend server)
    
**required**: no

**type**: string

**default**: http://localhost:8080

### `files`: 
    
**about**: Files to watch for changes (usually best to watch the compiled files if possible so we don't inject/reload too early)
    
**required**: yes

**type**: string | array (of globs)
    
**default**: null

### `openBrowser`: 
    
**about**: Open browser to proxy server on start?
    
**required**: no

**type**: boolean
    
**default**: true


CHANGELOG.md
# Changelog


<h2 class="ReleaseVersion">0.9.4</h2>

<p class="ReleaseDate">
  <time datetime="2021-07-01T13:40:40.334Z">07/01/2021</time>
</p>


### Updated
- Updated dependencies with npm


<h2 class="ReleaseVersion">0.9.3</h2>

<p class="ReleaseDate">
  <time datetime="2021-05-05T13:19:36.450Z">05/05/2021</time>
</p>


### Updated
- NPM dependencies
- Node version to 16


<h2 class="ReleaseVersion">0.9.2</h2>

<p class="ReleaseDate">
  <time datetime="2021-03-29T14:13:56.580Z">03/29/2021</time>
</p>


### Fixed
- add longer timeout to beforeAll in test suite for slower machines


<h2 class="ReleaseVersion">0.9.1</h2>

<p class="ReleaseDate">
  <time datetime="2021-03-29T14:09:12.628Z">03/29/2021</time>
</p>


### Updated
- Dependencies updated


<h2 class="ReleaseVersion">0.9.0</h2>

<p class="ReleaseDate">
  <time datetime="2020-10-21T14:44:10.118Z">10/21/2020</time>
</p>


### Updated
- Bumped deps


<h2 class="ReleaseVersion">0.8.2</h2>

<p class="ReleaseDate">
  <time datetime="2020-08-13T13:03:13.646Z">08/13/2020</time>
</p>


### Updated
- bump dependencies


<h2 class="ReleaseVersion">0.8.1</h2>

<p class="ReleaseDate">
  <time datetime="2020-07-27T12:56:46.704Z">07/27/2020</time>
</p>


### Updated
- Add a queue system to prevent multiple refreshes within _n_ ms of each other


<h2 class="ReleaseVersion">0.8.0</h2>

<p class="ReleaseDate">
  <time datetime="2020-07-21T23:25:48.501Z">07/21/2020</time>
</p>


### New
- added openBrowser option (defaults to true) that will open default browser to the proxied server address on start


<h2 class="ReleaseVersion">0.7.2</h2>

<p class="ReleaseDate">
  <time datetime="2020-07-21T19:47:52.414Z">07/21/2020</time>
</p>


### Fixed
- remove dead code


<h2 class="ReleaseVersion">0.7.1</h2>

<p class="ReleaseDate">
  <time datetime="2020-07-21T19:44:08.785Z">07/21/2020</time>
</p>


### Fixed
- fix pass through of data that is not HTML in node responses


<h2 class="ReleaseVersion">0.7.0</h2>

<p class="ReleaseDate">
  <time datetime="2020-07-19T13:31:23.436Z">07/19/2020</time>
</p>


### Updated
- Drop harmon/trumpet in favor of simple middleware


<h2 class="ReleaseVersion">0.6.1</h2>

<p class="ReleaseDate">
  <time datetime="2020-07-16T12:08:28.492Z">07/16/2020</time>
</p>


### Fixed
- Logger time format


<h2 class="ReleaseVersion">0.6.0</h2>

<p class="ReleaseDate">
  <time datetime="2020-07-15T21:01:34.075Z">07/15/2020</time>
</p>


### New
- Added Jest
- Added tests with playwright
- CLI updates (added logs option)

### Updated
- Minor refactor


src/anubis.js
#!/usr/bin/env node
const yargs = require('yargs')
const chalk = require('chalk')
const chokidar = require('chokidar')
const httpProxy = require('http-proxy')
const http = require('http')
const socketio = require('socket.io')
const connect = require('connect')
const harmon = require('harmon')
const serveStatic = require('serve-static')
const finalhandler = require('finalhandler')
const path = require('path')
const { log } = console

const defaults = {
  files: null,
  target: 'http://localhost:8080',
  port: 3000
}

const missingOptionsMessage = 'Anubis was asked to watch nothing! Anubis must be given an array or glob of files to be watched with the --files option'

const scriptsToInject = (port) => {
  return `
<!-- injected via Anubis -->
<script src="/socket.io/socket.io.js"></script>
<script src="http://localhost:${port}/anubis-client.js"></script>
<!-- /injected via Anubis -->
`
}

const Anubis = (userOptions) => {
  const opts = Object.assign({}, defaults, userOptions)

  if (!opts.files) {
    log(chalk.redBright(missingOptionsMessage))
    throw new Error('Missing required options! (files)')
  }

  const logger = {
    onStart () {
      log(
        chalk.green('\nAnubis is watching 👀'),
        chalk.blue(`\nhttp://localhost:${opts.port} 🆙\n`)
      )
    },
    onClientConnect (socket) {
      log(
        this.timeStamp() +
        chalk.magenta('[⚭ browser connected]') +
        chalk.blue(` ${socket.handshake.headers.host}`)
      )
    },
    onFileUpdated (event, filePath) {
      const time = this.timeStamp()
      const message = filePath.indexOf('.css') > -1 ? 'Injecting CSS!' : 'Reloading browser!'
      log(
        time +
        chalk.magenta(`[${event}] `) +
        chalk.green(`${filePath}`)
      )
      log(
        time +
        chalk.cyan(` ↳  ${message}`)
      )
    },
    timeStamp () {
      const now = new Date()
      return chalk.dim(
        `[${now.getHours()}:${now.getMinutes()}:${now.getSeconds()}] `
      )
    }
  }

  const injectClient = (node) => {
    const read = node.createReadStream()
    const write = node.createWriteStream({ outer: false })
    read.pipe(write, { end: false })
    read.on('end', () => {
      write.end(scriptsToInject(opts.port))
    })
  }

  const createServer = () => {
    const app = connect()
    const proxied = httpProxy.createProxyServer({
      target: opts.target
    })
    const server = http.createServer(app)
    const io = socketio(server)
    app.use(harmon([], [{
      query: 'body',
      func: injectClient
    }]))
    app.use((req, res, next) => {
      if (req.url !== '/anubis-client.js') proxied.web(req, res)
      else {
        const serve = serveStatic(path.join(__dirname))
        serve(req, res, finalhandler(req, res))
      }
    })
    server.listen(opts.port)

    return { io }
  }

  return () => {
    logger.onStart()
    const { io } = createServer()

    io.on('connection', logger.onClientConnect.bind(logger))

    chokidar
      .watch(opts.files)
      .on('all', (event, filePath) => {
        logger.onFileUpdated(event, filePath)
        io.emit('filesUpdated', filePath)
      })
  }
}

yargs
  .command({
    command: ['start', '$0'],
    describe: 'Start Anubis',
    builder: {
      files: {
        alias: 'f',
        describe: 'Files to watch'
      },
      target: {
        alias: 't',
        describe: 'URL to proxy',
        default: 'http://localhost:8080',
        type: 'string'
      },
      port: {
        alias: 'p',
        describe: 'Port to proxy to',
        default: 3000,
        type: 'number'
      }
    },
    handler (argv) {
      Anubis({
        files: argv.files,
        target: argv.target,
        port: argv.port
      })()
    }
  })
  .demandOption(
    ['files'],
    missingOptionsMessage
  )
  .parse()

module.exports = Anubis
